### PR TITLE
Adds tests and mode for dirichlet multinomial distribution

### PR DESCRIPTION
**Pull request by morganstrom**
The code is almost identical to that for the multinomial distribution (PR #5201)

**Thank your for opening a PR!**

Depending on what your PR does, here are a few things you might want to address in the description:
+ [ ] what are the (breaking) changes that this PR makes?
+ [ ] important background, or details about the implementation
+ [ ] are the changes—especially new features—covered by tests and docstrings?
+ [ ] [linting/style checks have been run](https://github.com/pymc-devs/pymc3/wiki/PyMC3-Python-Code-Style)
+ [ ] [consider adding/updating relevant example notebooks](https://github.com/pymc-devs/pymc-examples)
+ [ ] right before it's ready to merge, mention the PR in the RELEASE-NOTES.md

### CODE REVIEW

File: pymc/distributions/multivariate.py [hunk 1]
```diff
@@ -632,6 +628,22 @@ def dist(cls, n, a, *args, **kwargs):
 
         return super().dist([n, a], **kwargs)
 
+    def get_moment(rv, size, n, a):
+        p = a / at.sum(a, axis=-1, keepdims=True)
+        if a.ndim > 1:
+            n = at.shape_padright(n)
+        if (a.ndim == 1) & (n.ndim > 0):
```
Comment (ricardoV94) [inline pymc/distributions/multivariate.py]:
```suggestion
        if (a.ndim == 1) and (n.ndim > 0):
```

**Review by ricardoV94 — COMMENTED** [review]
I am not sure the shape handling is completely correct with the V4 version of the DM. You may want to check how the `rng_fn` is handling broadcasting. It might suffice to do this line: 

https://github.com/pymc-devs/pymc/blob/9d4691c4ceaa18dd50081db32f38fd1310359d48/pymc/distributions/multivariate.py#L582

This may also apply to the Multinomial distribution:
https://github.com/pymc-devs/pymc/blob/9d4691c4ceaa18dd50081db32f38fd1310359d48/pymc/distributions/multivariate.py#L468

**Comment by morganstrom** [comment]
> I am not sure the shape handling is completely correct with the V4 version of the DM. You may want to check how the `rng_fn` is handling broadcasting. It might suffice to do this line:
> 
> https://github.com/pymc-devs/pymc/blob/9d4691c4ceaa18dd50081db32f38fd1310359d48/pymc/distributions/multivariate.py#L582
> 
> This may also apply to the Multinomial distribution:
> 
> https://github.com/pymc-devs/pymc/blob/9d4691c4ceaa18dd50081db32f38fd1310359d48/pymc/distributions/multivariate.py#L468

Thanks, I wasn't aware of this function! I'm getting incompatible shapes when I try to use it however.. The code you refer to uses a constant (`cls.ndims_params`) that is set to [0, 1], which doesn't seem to work here.. 

https://github.com/pymc-devs/pymc/blob/9d4691c4ceaa18dd50081db32f38fd1310359d48/pymc/distributions/multivariate.py#L571

I think the issue might be that dimension 0 of `n` (in case of an array) is required to match dimension 1 of `a` - this is why we're using `at.shape_padright` for `n` and `at.shape_padleft` for `a` :  

https://github.com/pymc-devs/pymc/blob/9d4691c4ceaa18dd50081db32f38fd1310359d48/pymc/distributions/multivariate.py#L630

**Comment by ricardoV94** [comment]
> I think the issue might be that dimension 0 of `n` (in case of an array) is required to match dimension 1 of `a` - this is why we're using `at.shape_padright` for `n` and `at.shape_padleft` for `a` :
> 
> https://github.com/pymc-devs/pymc/blob/9d4691c4ceaa18dd50081db32f38fd1310359d48/pymc/distributions/multivariate.py#L630

That sounds like an old constraint that is no longer needed in V4. This works just fine on the random side, we might just need to check it also works on the logp:

```python
y = pm.DirichletMultinomial.dist(n=np.arange(1, 1+2*4).reshape(2, 4), a=np.ones(3))
y.eval()
```
```
array([[[1, 0, 0],
        [0, 2, 0],
        [0, 0, 3],
        [3, 0, 1]],
       [[1, 2, 2],
        [0, 5, 1],
        [0, 3, 4],
        [2, 0, 6]]])
```

**Comment by ricardoV94** [comment]
> > I think the issue might be that dimension 0 of `n` (in case of an array) is required to match dimension 1 of `a` - this is why we're using `at.shape_padright` for `n` and `at.shape_padleft` for `a` :
> > https://github.com/pymc-devs/pymc/blob/9d4691c4ceaa18dd50081db32f38fd1310359d48/pymc/distributions/multivariate.py#L630
> 
> That sounds like an old constraint that is no longer needed in V4. This works just fine on the random side, we might just need to check it also works on the logp:
> 
> ```python
> y = pm.DirichletMultinomial.dist(n=np.arange(1, 1+2*4).reshape(2, 4), a=np.ones(3))
> y.eval()
> ```
> 
> ```
> array([[[1, 0, 0],
>         [0, 2, 0],
>         [0, 0, 3],
>         [3, 0, 1]],
>        [[1, 2, 2],
>         [0, 5, 1],
>         [0, 3, 4],
>         [2, 0, 6]]])
> ```

**Comment by ricardoV94** [comment]
I am lifting the constraint of `n` in #5234

**Comment by ricardoV94** [comment]
@morganstrom in #5234 I removed the old constraint on the dimensionality of `n` and `p`. The `get_moment` should share the same flexibility, and we have to do the same for the `Multinoimal`. Do you want to incorporate those updates in this PR?

**Comment by twiecki** [comment]
@morganstrom Any update on this?

**Comment by morganstrom** [comment]
I'm sorry, have been meaning to work on this but haven't had the time or
energy. Perhaps somebody else can pick it up instead?

Den fre 7 jan. 2022 10:58Thomas Wiecki ***@***.***> skrev:

File: pymc/distributions/multivariate.py [hunk 2]
```diff
@@ -629,12 +629,7 @@ def dist(cls, n, a, *args, **kwargs):
         return super().dist([n, a], **kwargs)
 
     def get_moment(rv, size, n, a):
-        p = a / at.sum(a, axis=-1, keepdims=True)
-        if a.ndim > 1:
-            n = at.shape_padright(n)
-        if (a.ndim == 1) & (n.ndim > 0):
-            n = at.shape_padright(n)
-            p = at.shape_padleft(p)
+        p = a / at.sum(a, axis=-1)
```
Comment (morganstrom) [inline pymc/distributions/multivariate.py:686]: @ricardoV94 Here I'm trying to mirror the changes you made in #5234

File: pymc/tests/test_distributions_moments.py [hunk 3]
```diff
@@ -1171,28 +1171,10 @@ def normal_sim(rng, mu, sigma, size):
...
-            np.array([1, 10]),
-            None,
-            np.array([[1, 0, 0, 0], [2, 3, 3, 2]]),
-        ),
-        (
-            np.array([[25, 25, 25, 25], [26, 26, 26, 22]]),
-            np.array([1, 10]),
-            2,
-            np.full((2, 2, 4), [[1, 0, 0, 0], [2, 3, 3, 2]]),
+            np.array([[1, 0, 0, 0], [2, 3, 3, 2]]),     # Dim: 2 x 4
```
Comment (morganstrom) [inline pymc/tests/test_distributions_moments.py]: I'm getting an error here: `TypeError: Cannot convert Type TensorType(int64, matrix) (of Variable Elemwise{Cast{int64}}.0) into Type TensorType(int64, (False, True, False))`

Comment (ricardoV94) [reply pymc/tests/test_distributions_moments.py]: It seems moment is returning a 3d tensor?

Comment (ricardoV94) [reply pymc/tests/test_distributions_moments.py]: You can try to call `get_moment` method directly and evaluate the returned graph, in order to facilitate debugging.

Comment (morganstrom) [reply pymc/tests/test_distributions_moments.py]: When I do what you suggest (in a debug console), I get a 2d array - confusing!
```
get_moment(model["x"]).eval()
array([[1, 0, 0, 0],
       [2, 3, 3, 2]])
```

Comment (ricardoV94) [reply pymc/tests/test_distributions_moments.py]: Are you using the latest Aesara version? There was a bug with shapes/casting, that was fixed recently.

Comment (ricardoV94) [reply pymc/tests/test_distributions_moments.py]: Okay, so in that example the distribution shape is actually `(2, 1, 4)`, but the returned moment is `(2, 4)`. 
```python
pm.DirichletMultinomial.dist(
    a=np.array([[26, 26, 26, 22]]),  # Dim: 1 x 4
    n=np.array([[1], [10]]),  # Dim: 2 x 1
).eval().shape  # (2, 1, 4)
```

**Comment by ricardoV94** [comment]
@morganstrom I rebased from main, so make sure to pull from here before doing further work

File: pymc/tests/test_distributions_moments.py (hunk 3 above)
Comment (ricardoV94) [reply pymc/tests/test_distributions_moments.py]: I pushed some changes that I think make the mode consistent with the RV shape. Does it look sensible to you?

This was a great edge case you stumbled upon. We could very well have missed it

**Review by ricardoV94 — APPROVED** [review]
